assets/shiny-server.js

```diff
@@ -637,6 +333,311 @@
+    }
+
+    // Compute delay exponentially.
+    var interval;
+    var delay;
+    if (count < 0){
+      interval = 0;
+      delay = 0;
+    } else {
+      if (count < 10){
+        interval = 1000 * Math.pow(2, count);
+        interval = Math.min(15 * 1000, interval); // Max of 15s delay.
+      } else { 
+        // The interval may end up being configurable or changed, so we don't cut off
+        // exactly at 2^4, but don't bother computing really large powers.
+        interval = 15 * 1000;
+      }
+      delay = time - Date.now() + interval;
+    }
+
+    // If the next attempt would be after the session is due to expire, 
+    // schedule one last attempt to connect a couple seconds before the
+    // expiration.
+    if (Date.now() + delay > (expires - 2000)) {
+      delay = expires - Date.now() - 2000;
+    }
+    if (delay < 0){
+      // i.e. we're within 2 seconds of session expiration. Make
+      // one last connection attmpt but don't schedule any more.
+      debug('Final reconnection attempt');
+      return this.reconnect_p();
+    }
+
+    var self = this;
+    function doRecon(){
+      var startTime = Date.now();
+      self.reconnect_p()
+      .then(function(c){
+        // Able to reconnect.
+        def.resolve(c);
+      }, function(){
+        self.scheduleReconnect_p(startTime, count+1, expires)
+        .then(function(c){
+          def.resolve(c);
+        }, function(e){
+          def.reject(e);
+        });
+      });
+    }
+
+    var reconTimeout = null;
+    this.setReconnectDialog('<button id="ss-reconnect-btn" type="button" class="ss-dialog-button">Reconnect Now</button>Trouble connecting to server.');
+    $('#ss-reconnect-btn').click(function(){
+      debug('Trying to reconnect now.');
+      // Short-circuit the wait.
+      clearTimeout(reconTimeout);
+      doRecon();
+    });
+    debug('Setting countdown timer');
+    debug('Scheduling reconnect attempt for ' + delay + 'ms');
+    // Schedule the reconnect for some time in the future.
+    reconTimeout = setTimeout(doRecon, delay);
+
+    return def;
+  }
+  MultiplexClient.prototype.startReconnect = function(){
+    var self = this;
+
+    // Schedule the display of the disconnect window
+    self._disconnectTimer = setTimeout(function(){
+      debug('Displaying disconnect screen.');
+      $('body').addClass('ss-reconnecting');
+      $('<div id="ss-connect-dialog">'+self._dialogMsg+'<div class="ss-clearfix"></div></div><div id="ss-overlay"></div>').appendTo('body');
+    }, 3500);
+
+    var timeout = 15; 
+
+
+    // Attempt to reconnect immediately, then start scheduling.
+    if (!self._disconnected){
+      var time = Date.now();
+      self.scheduleReconnect_p(time, -1, time + 15 * 1000)
+      .fail(function(){
+        // We were not able to re/connect
+        self._disconnected = true;
+        self._doClose();
+      });
+    } else {
+      self._disconnected = true;
```

Comment:
nit: `_disconnected` is already set and we have done nothing to change our connection state.